### ISSUE TITLE
Automatic password reset

### ISSUE THREAD

**Issue by megalodonatron** [qiime/qiime_web_app#565]
It would be nice to have a systematic password reset mechanism so if people forget their passwords we don't have to deal with it in the help account.

**Comment by ElDeveloper** [comment]
There's a change password button in the sidebar but you have to be logged in to be able to see this. I guess we probably want something that sends them an e-mail with their new password?

**Comment by adamrp** [comment]
Is this worth developer effort?  I don't think I have ever had to do this
for anyone.

Adam

**Comment by megalodonatron** [comment]
I've answered many "I forgot my password" emails every time that I've been AG help master, but if no one else has that experience then don't worry about it.

**Comment by ElDeveloper** [comment]
I'm not sure if this is "worth the developer effort", that's a rather delicate question, though I do agree with Meg that this is a question that I've seen. Additionally it seems counterintuitive to have to know your password to be able to reset your password, or at least it seems to me because I commonly do this with my accounts in different websites.

**Comment by adamrp** [comment]
Oh, sorry, I didn't realize this was for the American Gut, thought it was
just the web app in general.

Note that the "change password" link is not intended to be a "reset
password" pipeline.  Having an automated system would be awesome.  Would
likely involve a new database table with timestamp, email address, and
shared secret.  In fact, it would be cool to do this for the kit
verification code, too, since people are  constantly saying they didn't get
that email.

Adam

**Comment by wasade** [comment]
I agree that the cumulative sum responding to forgotten password requests
is quite high. Yoshiki or Adam, do either of you have an estimate on the
amount of developer effort necessary to expose this functionality?

**Comment by adamrp** [comment]
Can get a rudimentary system in place, but what is the full list of
features we want?  Here's what I'm thinking:
- On index.psp, add link for "reset password"
- Small form for "Enter your email address here, and we will send you a
  link to reset your password"
- Another small form (after the user has clicked the link) to enter a new
  password

And some thoughts:
- Need to add columns to ag_kit table: "reset code" (or equivalent) and
  "reset_timestamp"
- Need to add reset_password.psp page
- When the user clicks "reset password" on index.psp, a small form is
  presented (reset_password.psp) and a random string is added as the
  reset_code, and the timestamp is recorded
- The link we send via email could be for a GET URL (e.g.,
  www.microbio.me/americangut/password.psp?q=############) where the #### is
  the random string from above
- If the page is visited (with q=that_same_random_string) within a certain
  amount of time, then the form is displayed to change the password.  In all
  other cases, the user is redirected to index.psp

How does that sound?

**Comment by wasade** [comment]
Possible to piggy back off the existing system for the qiime webapp?

That sounds reasonable. Do you by chance have a rough estimate of developer
hours needed?

**Comment by adamrp** [comment]
Oo, yeah, just took a look at that and that's pretty much what I am
describing.  I don't think we can rip it directly, but probably with just a
little modification.  I won't know about how much time it will take until I
have a look at the code that's currently in place.

**Comment by teravest** [comment]
I think we need something to retrieve kit ids by email, as well as something to retrieve passwords. I created a stored procedure on the test DB to retrieve kit ID by email.  Does anyone know if we have an IRB issue with creating a form to  send a kit ID by email based on email?

**Comment by adamrp** [comment]
IRB questions should be directed to Gail, so shoot her an email.

**Comment by ElDeveloper** [comment]
cc. @ackermang.

**Comment by wasade** [comment]
I think we want to allow them to change passwords, not retrieve. Agree that
we need @ackermang to comment on the kit ID

**Comment by teravest** [comment]
I agree on changing password. Sorry I was unclear in my last message.  I
sent @ackermang an email asking about kit I'd retrieval.
